spurious generic crashes gcc

I made a mistake and included a `generic [T: Free]` on a non-generic function; instead of an Austral compiler error, it crashed gcc:

```
module body MutateVar is
    generic [T: Free] -- spurious
    function mutate_int(var: Int32): Unit is
        return nil;
    end;

    function main(): ExitCode is
        mutate_int(5);
        return ExitSuccess();
    end;
end module body.
```

```
Compiler call tree printed to calltree.html
Error:
  Title: Generic Error
  Module:
    [unknown]
  Location:
    [no span available]
  Description:
    Error when running C compiler.
      Command: cc /tmp/austral_17dab9.c -fwrapv -lm -o mine/mutate_var
      Exit code: 1
      Standard output:
    
      Standard error:
    /tmp/austral_17dab9.c: In function 'decl_219':
    /tmp/austral_17dab9.c:3929:5: warning: implicit declaration of function 'decl_218'; did you mean 'decl_219'? [-Wimplicit-function-declaration]
     3929 |     decl_218(((au_int32_t) 5));
          |     ^~~~~~~~
          |     decl_219
    /usr/bin/ld: /tmp/ccLzEeL8.o: in function `decl_219':
    austral_17dab9.c:(.text+0x3fb2): undefined reference to `decl_218'
    collect2: error: ld returned 1 exit status
    
  Code:
    [no span available]
```

--- a/lib/ExtractionPass.ml
+++ b/lib/ExtractionPass.ml
@@ -25,6 +25,7 @@ open TypeParameters
 open TypeClasses
 open Region
 open BuiltIn
+open TypeVarSet
 open Util
 open Error
 
@@ -114,7 +115,14 @@ module Errors = struct
     austral_raise DeclarationError [
       Text "Cannot find typeclass with the name ";
       Code (ident_string name)
-    ]
+      ]
+
+  let typaram_not_in_signature (tp: type_parameter) =
+    austral_raise DeclarationError [
+        Text "The type parameter ";
+        Code (ident_string (typaram_name tp));
+        Text " does not appear anywhere in the signature of the function."
+      ]
 end
 
 let check_slots_are_free (name: identifier) (slots: typed_slot list): unit =
@@ -154,8 +162,22 @@ let check_typarams_are_free (name: identifier) (typarams: typarams): unit =
   List.iter (fun tp -> check_universe (typaram_universe tp)) (typarams_as_list typarams)
 
 let check_all_type_parameters_appear_in_signature (typarams: typarams) (params: value_parameter list) (rt: ty): unit =
-  let _ = (typarams, params, rt) in
-  ()
+  (* Get the type variables from the value parameter list. *)
+  let param_tyvars: TypeVarSet.t list = List.map (fun (ValueParameter (_, ty)) -> type_variables ty) params in
+  let param_tyvars: TypeVarSet.t = List.fold_left TypeVarSet.union TypeVarSet.empty param_tyvars in
+  (* Get the type variables from the return type. *)
+  let rt_tyvars: TypeVarSet.t = type_variables rt in
+  (* Merge them to get the type variables in the signature. *)
+  let signature_tyvars: TypeVarSet.t = TypeVarSet.union param_tyvars rt_tyvars in
+  (* Go through each of the type parameters, and assert that it is in `singature_tyvars`. *)
+  let check_typaram (tp: type_parameter): unit =
+    let tv: type_var = typaram_to_tyvar tp in
+    if TypeVarSet.mem tv signature_tyvars then
+      ()
+    else
+      Errors.typaram_not_in_signature tp
+  in
+  List.iter check_typaram (typarams_as_list typarams)
 
 let rec extract_type_signatures (CombinedModule { decls; _ }): type_signature list =
   List.filter_map extract_type_signatures' decls

--- a/lib/ExtractionPass.ml
+++ b/lib/ExtractionPass.ml
@@ -153,6 +153,10 @@ let check_typarams_are_free (name: identifier) (typarams: typarams): unit =
   in
   List.iter (fun tp -> check_universe (typaram_universe tp)) (typarams_as_list typarams)
 
+let check_all_type_parameters_appear_in_signature (typarams: typarams) (params: value_parameter list) (rt: ty): unit =
+  let _ = (typarams, params, rt) in
+  ()
+
 let rec extract_type_signatures (CombinedModule { decls; _ }): type_signature list =
   List.filter_map extract_type_signatures' decls
 
@@ -377,6 +381,10 @@ and extract_definition (env: env) (mod_id: mod_id) (mn: module_name) (local_type
             ()
        | None ->
           ()
+     in
+     let _ =
+       (* Check: all type parameters must appear in the type signature. *)
+       check_all_type_parameters_appear_in_signature typarams value_params rt
      in
      let fn_input: function_input = {
          mod_id = mod_id;

--- a/test-programs/suites/011-bugs/github-424/README.md
+++ b/test-programs/suites/011-bugs/github-424/README.md
@@ -1,0 +1,1 @@
+Regression test for [GitHub issue #424](https://github.com/austral/austral/issues/424).

--- a/test-programs/suites/011-bugs/github-424/Test.aum
+++ b/test-programs/suites/011-bugs/github-424/Test.aum
@@ -1,0 +1,11 @@
+module body Test is
+    generic [T: Free] -- spurious
+    function mutate_int(var: Int32): Unit is
+        return nil;
+    end;
+
+    function main(): ExitCode is
+        mutate_int(5);
+        return ExitSuccess();
+    end;
+end module body.

--- a/test-programs/suites/011-bugs/github-424/austral-stderr.txt
+++ b/test-programs/suites/011-bugs/github-424/austral-stderr.txt
@@ -1,0 +1,10 @@
+Error:
+  Title: Declaration Error
+  Module:
+    Test
+  Location:
+    [no span available]
+  Description:
+    The type parameter `T` does not appear anywhere in the signature of the function.
+  Code:
+    [no span available]

